### PR TITLE
[Enhancement] Avoid meaningless tablet repair scheduling

### PR DESCRIPTION
**Pull request by nshangyiming**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #7658

## Problem Summary(Required) ：

It's explained in the issue #7658 in detail.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 10 (20.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Review by imay — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletChecker.java [hunk 1]
```diff
@@ -189,6 +189,24 @@ protected void runAfterCatalogReady() {
...
+     * In order to avoid meaningless repair schedule, for task that need to
+     * choose a source replica to clone from, we check that whether we can
+     * find a healthy source replica before send it to pending queue.
+     * <p>
+     * If we don't do this check, the task will go into the pending queue and
+     * get deleted when no healthy source replica found, and then it will be added
+     * to the queue again in the next `TabletChecker` round.
+     */
+    private boolean tryChooseSrcBeforeSchedule(TabletSchedCtx tabletCtx) {
+        if (tabletCtx.needCloneFromSource() &&
```
[the remark this answers is not in the document]
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/clone/TabletChecker.java:202]: good idea

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Review by imay — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/clone/TabletSchedCtx.java [hunk 2]
```diff
@@ -495,13 +495,16 @@ public boolean isDestPathResourceHold() {
...
-        /*
-         * get all candidate source replicas
-         * 1. source replica should be healthy.
-         * 2. slot of this source replica is available.
-         */
+    public boolean needCloneFromSource() {
+        return tabletStatus == TabletStatus.REPLICA_MISSING ||
+                tabletStatus == TabletStatus.VERSION_INCOMPLETE ||
+                tabletStatus == TabletStatus.REPLICA_RELOCATING ||
+                tabletStatus == TabletStatus.FORCE_REDUNDANT ||
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/clone/TabletSchedCtx.java:502]: `FORCE_REDUNDANT` does not need to check source replica health.

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/clone/TabletSchedCtx.java:502]: Yes, `FORCE_REDUNDANT` is just like `REDUNDANT`, only delete one replica, but in force mode, and let the repair to be done later. I will remove this condition.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wyb** [comment]
run starrocks_fe_unittest

**Review by wyb — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by nshangyiming** [comment]
@mergify backport branch-2.3

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 2 / 7 (28.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/clone/TabletChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/clone/TabletSchedCtx.java|2|3|66.67%|

**Pull request merged**